File: chapter20-boot/src/main/java/com/apress/prospring6/twenty/boot/service/SingerService.java
package com.apress.prospring6.twenty.boot.service;

import com.apress.prospring6.twenty.boot.model.Singer;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.validation.Errors;
import org.springframework.validation.ValidationUtils;
import org.springframework.validation.Validator;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

public interface SingerService {

    Flux<Singer> findAll();
    Mono<Singer> findById(Long id);
    Mono<Singer> findByFirstNameAndLastName(String firstName, String lastName);
    Flux<Singer> findByFirstName(String firstName);
    Mono<Singer> save(Singer singer);
    Mono<Singer> update(Long id, Singer actorMon);
    Mono<Void> delete(Long id);

    Flux<Singer> findByCriteriaDto(CriteriaDto criteria);

    @Getter
    @Setter
    @NoArgsConstructor
    class CriteriaDto {
        private String fieldName;
        private String fieldValue;
    }

    class CriteriaValidator implements Validator {
        @Override
        public boolean supports(Class<?> clazz) {
            return (CriteriaDto.class).isAssignableFrom(clazz);
        }

        @Override
        public void validate(Object target, Errors errors) {
            ValidationUtils.rejectIfEmpty(errors, "fieldName", "required", new Object[]{"fieldName"}, "Field Name is required!");
            ValidationUtils.rejectIfEmpty(errors, "fieldValue", "required", new Object[]{"fieldValue"}, "Field Value is required!");
        }
    }

    enum FieldGroup {
        FIRSTNAME,
        LASTNAME,
        BIRTHDATE;

        public static FieldGroup getField(String field) {
            return FieldGroup.valueOf(field.toUpperCase());
        }
    }
}

File: chapter16-kotlin/src/main/kotlin/com/apress/prospring6/sixteen/kotlin/boot/SingerService.kt
package com.apress.prospring6.sixteen.kotlin.boot

import jakarta.transaction.Transactional
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.stereotype.Service
import kotlin.jvm.Throws

@Transactional
@Service
class SingerService(private val repository: SingerRepository) {

    @Throws(NotFoundException::class)
    fun findAll(): List<Singer> {
        val singers: List<Singer> = repository.findAll() as List<Singer>
        if (singers.isEmpty()) {
                throw NotFoundException(Singer::javaClass.name)
        }
        return singers
    }

    @Throws(NotFoundException::class)
    fun findById(id: Long?): Singer? {
        return id?.let { repository.findById(id).orElseThrow { NotFoundException(Singer::javaClass.name, id)}}
    }

    @Throws(DataIntegrityViolationException::class)
    fun save(singer: Singer?): Singer? {
        return singer?.let { repository.save(it) }
    }

    @Throws(NotFoundException::class, DataIntegrityViolationException::class)
    fun update(id: Long?, singer: Singer): Singer? {
        return id?.let { repository.findById(id).map { update(it, singer, repository)}
            .orElseThrow {NotFoundException(Singer::javaClass.name, id)}}
    }

    @Throws(NotFoundException::class)
    fun delete(id: Long?) {
        id?.let { repository.findById(id).orElseThrow { NotFoundException(Singer::javaClass.name, id)}}
        id?.let { repository.deleteById(id)}
    }

    fun update(it: Singer, singer: Singer, repository: SingerRepository): Singer {
        it.firstName = singer.firstName
        it.lastName = singer.lastName
        it.birthDate = singer.birthDate

        return repository.save(it)
    }
}